python/cuml/dask/common/comms.py

```diff
@@ -0,0 +1,495 @@
+    def init_nccl(self):
+        """
+        Use nccl-py to initialize ncclComm_t on each worker and
+        store the futures for this instance.
+        """
+        self.uniqueId = nccl.get_unique_id()
+
+        for worker, idx in zip(self.worker_addresses,
+                               range(len(self.workers))):
+            self.client.run(_func_store_initial_state,
+                            len(self.workers),
+                            self.sessionId,
+                            self.uniqueId,
+                            idx,
+                            workers=[worker])
+
+        self.client.run(_func_init_nccl,
+                        self.sessionId,
+                        wait=False)
+
+        self.block_for_init("nccl")
+
+        self.nccl_initialized = True
+
+    def init_ucp(self):
+        """
+        Use ucx-py to initialize ucp endpoints so that every
+        worker can communicate, point-to-point, with every other worker
+        """
+        self.create_ucp_listeners()
+        self.ucp_create_endpoints()
+
+        self.ucx_initialized = True
+
+    def init(self):
+        """
+        Initializes the underlying comms. NCCL is required but
+        UCX is only initialized if `comms_p2p == True`
+        """
+        self.init_nccl()
+
+        if self.comms_p2p:
+            self.init_ucp()
+
+            self.client.run(_func_build_handle_p2p,
+                            self.sessionId,
+                            wait=False)
+
+        else:
+            self.client.run(_func_build_handle,
+                            self.sessionId,
+                            wait=False)
+
+        self.block_for_init("handle")
+
+        _set_global_comms(self)
+
+    def ucp_create_endpoints(self):
+        """
+        Creates UCX endpoints for each worker in the Dask cluster and
+        connects them to every other worker.
+        """
+        worker_info = self.worker_info()
+
+        [self.client.run(_func_ucp_create_endpoints, self.sessionId,
+                         worker_info, workers=[w],
+                         wait=False)
+         for w in self.worker_addresses]
+
+        self.block_for_init("ucp_eps")
+
+    def destroy_nccl(self):
+        """
+        Destroys all NCCL communicators on workers
+        """
+        self.client.run(_func_destroy_nccl, self.sessionId, wait=True)
+
+    def destroy_eps(self):
+        """
+        Destroys all UCP endpoints on all workers
+        """
+        self.client.run(_func_destroy_ep,
+                        self.sessionId,
+                        wait=True)
+
+    def destroy_ucp(self):
+        """
+        Stops initialized UCP endpoints and listers on the Dask workers
+        """
+        self.destroy_eps()
+        self.stop_ucp_listeners()
+
+    def destroy(self):
+        """
+        Shuts down initialized comms and cleans up resources.
+        """
+        if self.comms_p2p:
+            self.destroy_ucp()
+
+        self.destroy_nccl()
```

Comment:
should this also call `destroy_eps`?

Reply:
No, `destroy_ucp()` calls `destroy_eps()`